File: killrvideo-service-users/src/test/java/com/killrvideo/service/user/grpc/UserManagementServiceGrpcTest.java
package com.killrvideo.service.user.grpc;

import com.killrvideo.messaging.dao.MessagingDao;
import com.killrvideo.service.user.dto.User;
import com.killrvideo.service.user.dto.UserCredentials;
import com.killrvideo.service.user.repository.UserRepository;
import com.killrvideo.utils.GrpcMappingUtils;
import io.grpc.stub.StreamObserver;
import killrvideo.user_management.UserManagementServiceOuterClass.*;
import killrvideo.user_management.events.UserManagementEvents;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;

import static com.killrvideo.utils.HashUtils.hashPassword;
import static java.util.Collections.singletonList;
import static org.mockito.Mockito.*;

@SuppressWarnings("unchecked")
class UserManagementServiceGrpcTest {
    @InjectMocks private UserManagementServiceGrpc service;
    @Mock
    private UserRepository userRepository;
    @Mock
    private MessagingDao messagingDao;
    @Mock
    private UserManagementServiceGrpcValidator validator;
    @Mock
    private UserManagementServiceGrpcMapper mapper;
    private AutoCloseable closeable;

    @BeforeEach
    public void openMocks() {
        closeable = MockitoAnnotations.openMocks(this);
    }

    @AfterEach
    public void releaseMocks() throws Exception {
        closeable.close();
    }

    @Test
    void testCreateUserWithValidationFailure() {
        CreateUserRequest grpcReq = CreateUserRequest.getDefaultInstance();
        StreamObserver<CreateUserResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_createUser(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.createUser(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testCreateUserWithCreateFailure() {
        CreateUserRequest grpcReq = createUserRequest("passwd", "joe@gmail.com");
        StreamObserver<CreateUserResponse> grpcResObserver = mock(StreamObserver.class);

        User user = mock(User.class);
        when(this.mapper.mapUserRequest2User(grpcReq)).thenReturn(user);

        doNothing().when(this.validator).validateGrpcRequest_createUser(any(), any());
        when(userRepository.createUserAsync(any(), any())).thenReturn(
            CompletableFuture.failedFuture(new Exception())
        );

        service.createUser(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testCreateUser() {
        CreateUserRequest grpcReq = createUserRequest("passwd", "joe@gmail.com");
        StreamObserver<CreateUserResponse> grpcResObserver = mock(StreamObserver.class);

        User user = mock(User.class);
        when(this.mapper.mapUserRequest2User(grpcReq)).thenReturn(user);

        UserManagementEvents.UserCreated event = UserManagementEvents.UserCreated.getDefaultInstance();
        when(this.mapper.createUserCreatedEvent(user)).thenReturn(event);

        doNothing().when(this.validator).validateGrpcRequest_createUser(any(), any());
        when(userRepository.createUserAsync(any(), any())).thenReturn(
                CompletableFuture.completedFuture(null)
        );
        when(this.messagingDao.sendEvent(any(), any())).thenReturn(
            CompletableFuture.completedFuture(null)
        );

        service.createUser(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testVerifyCredentialsWithValidationFailure() {
        VerifyCredentialsRequest grpcReq = VerifyCredentialsRequest.getDefaultInstance();
        StreamObserver<VerifyCredentialsResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_VerifyCredentials(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.verifyCredentials(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testVerifyCredentialsWithQueryFailure() {
        VerifyCredentialsRequest grpcReq = verifyCredentialsRequest("joe@gmail.com", "passwd");
        StreamObserver<VerifyCredentialsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_VerifyCredentials(any(), any());

        when(userRepository.getUserCredentialAsync(any())).thenReturn(
          CompletableFuture.failedFuture(new Exception())
        );

        service.verifyCredentials(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testVerifyCredentialsWithPasswdMismatch() {
        VerifyCredentialsRequest grpcReq = verifyCredentialsRequest("joe@gmail.com", "passwd");
        StreamObserver<VerifyCredentialsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_VerifyCredentials(any(), any());

        UserCredentials credentials = userCredentials("joe@gmail.com", "xxx");
        when(userRepository.getUserCredentialAsync(any())).thenReturn(
                CompletableFuture.completedFuture(credentials)
        );

        service.verifyCredentials(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testVerifyCredentials() {
        VerifyCredentialsRequest grpcReq = verifyCredentialsRequest("joe@gmail.com", "passwd");
        StreamObserver<VerifyCredentialsResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_VerifyCredentials(any(), any());

        UserCredentials credentials = userCredentials("joe@gmail.com", hashPassword("passwd"));
        when(userRepository.getUserCredentialAsync(any())).thenReturn(
                CompletableFuture.completedFuture(credentials)
        );

        service.verifyCredentials(grpcReq, grpcResObserver);
        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetUserProfileWithValidationFailure() {
        GetUserProfileRequest grpcReq = GetUserProfileRequest.getDefaultInstance();
        StreamObserver<GetUserProfileResponse> grpcResObserver = mock(StreamObserver.class);

        doThrow(new IllegalArgumentException()).when(this.validator)
                .validateGrpcRequest_getUserProfile(any(), any());

        Assertions.assertThrows(IllegalArgumentException.class, () ->
                service.getUserProfile(grpcReq, grpcResObserver)
        );
    }

    @Test
    void testGetUserProfileWithEmptyUserIds() {
        GetUserProfileRequest grpcReq = getUserProfileRequest();
        StreamObserver<GetUserProfileResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getUserProfile(any(), any());
        service.getUserProfile(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @Test
    void testGetUserProfileWithQueryFailure() {
        UUID userId = UUID.randomUUID();
        GetUserProfileRequest grpcReq = getUserProfileRequest(userId);
        StreamObserver<GetUserProfileResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getUserProfile(any(), any());
        when(this.mapper.mapToListOfUserid(any())).thenReturn(singletonList(userId));

        when(this.userRepository.getUserProfilesAsync(any())).thenReturn(
          CompletableFuture.failedFuture(new Exception())
        );
        service.getUserProfile(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(1)).onError(any());
        verify(grpcResObserver, times(0)).onNext(any());
        verify(grpcResObserver, times(0)).onCompleted();
    }

    @Test
    void testGetUserProfile() {
        UUID userId = UUID.randomUUID();
        GetUserProfileRequest grpcReq = getUserProfileRequest(userId);
        StreamObserver<GetUserProfileResponse> grpcResObserver = mock(StreamObserver.class);

        doNothing().when(this.validator).validateGrpcRequest_getUserProfile(any(), any());
        when(this.mapper.mapToListOfUserid(any())).thenReturn(singletonList(userId));

        User user = mock(User.class);
        List<User> users = singletonList(user);

        GetUserProfileResponse response = GetUserProfileResponse.getDefaultInstance();
        when(this.mapper.buildGetUserProfileResponse(users)).thenReturn(response);

        when(this.userRepository.getUserProfilesAsync(any())).thenReturn(
                CompletableFuture.completedFuture(users)
        );
        service.getUserProfile(grpcReq, grpcResObserver);

        verify(grpcResObserver, times(0)).onError(any());
        verify(grpcResObserver, times(1)).onNext(any());
        verify(grpcResObserver, times(1)).onCompleted();
    }

    @SuppressWarnings("SameParameterValue")
    private CreateUserRequest createUserRequest(String passwd, String email) {
        return CreateUserRequest.newBuilder()
                .setPassword(passwd)
                .setEmail(email)
                .build();
    }

    @SuppressWarnings("SameParameterValue")
    private VerifyCredentialsRequest verifyCredentialsRequest(String email, String passwd) {
        return VerifyCredentialsRequest.newBuilder()
                .setEmail(email)
                .setPassword(passwd)
                .build();
    }

    @SuppressWarnings("SameParameterValue")
    private UserCredentials userCredentials(String email, String passwd) {
        User user = mock(User.class);
        when(user.getEmail()).thenReturn(email);
        return UserCredentials.from(user, passwd);
    }

    private GetUserProfileRequest getUserProfileRequest(UUID... userIds) {
        return GetUserProfileRequest.newBuilder()
                .addAllUserIds(
                        Arrays.stream(userIds).map(GrpcMappingUtils::uuidToUuid)
                                .collect(Collectors.toList())
                )
                .build();
    }
}

File: killrvideo-service-users/src/test/kotlin/com/killrvideo/service/user/grpc/UserManagementServiceGrpcTest.kt
package com.killrvideo.service.user.grpc

import com.killrvideo.messaging.dao.MessagingDao
import com.killrvideo.service.user.dto.User
import com.killrvideo.service.user.dto.UserCredentials
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import com.killrvideo.utils.HashUtils
import io.mockk.*
import io.mockk.impl.annotations.InjectMockKs
import io.mockk.impl.annotations.MockK
import killrvideo.user_management.*
import killrvideo.user_management.UserManagementServiceOuterClass.UserProfile
import killrvideo.user_management.events.UserManagementEvents.UserCreated
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.util.*
import java.util.concurrent.CompletableFuture

internal class UserManagementServiceGrpcTest {
    @InjectMockKs
    private lateinit var service: UserManagementServiceGrpc

    @MockK
    private lateinit var userService: UserManagementService

    @MockK
    private lateinit var messagingDao: MessagingDao

    @MockK
    private lateinit var validator: UserManagementServiceGrpcValidator

    @MockK
    private lateinit var mapper: UserManagementServiceGrpcMapper

    private val topicUserCreated = "topic-kv-userCreation"
    val serviceKey = "UserManagementService"

    @BeforeEach
    fun setUp() = MockKAnnotations.init(this, relaxUnitFun = true)

    @Test
    fun testCreateUserWithValidationFailure() {
        val request = createUserRequest {}

        every {
            validator.validateGrpcRequest_createUser(any())
        } throws IllegalArgumentException()

        assertThrows<IllegalArgumentException> {
            runBlocking {
                service.createUser(request)
            }
        }
    }

    @Test
    fun testCreateUserWithCreateFailure() {
        val userid = UUID.randomUUID()
        val request = createUserRequest { password = "passwd"; email = "joe@gmail.com"; userId=uuidToUuid(userid) }
        every { validator.validateGrpcRequest_createUser(any()) } just Runs

        coEvery { userService.createUser(any(), any()) } returns Result.failure(Exception())

        assertThrows<Exception> {
            runBlocking {
                service.createUser(request)
            }
        }
    }

    @Test
    fun testCreateUser() {
        val userid = UUID.randomUUID()
        val request = createUserRequest { password = "passwd"; email = "joe@gmail.com"; userId=uuidToUuid(userid) }
        val user = mockk<User>()
        every { validator.validateGrpcRequest_createUser(any()) } just Runs

        val event = UserCreated.getDefaultInstance()
        every { mapper.createUserCreatedEvent(any()) } returns event

        coEvery { userService.createUser(any(), any()) } returns Result.success(true)
        every { messagingDao.sendEvent(any(), any()) } returns CompletableFuture.completedFuture(null)

        runBlocking {
            service.createUser(request)
        }

        coVerify {
            userService.createUser(any(), any())
        }

        verify(exactly = 1) {
            messagingDao.sendEvent(any(), any())
        }
    }

    @Test
    fun testVerifyCredentialsWithValidationFailure() {
        val request = verifyCredentialsRequest {}
        every { validator.validateGrpcRequest_VerifyCredentials(any()) } throws IllegalArgumentException()

        assertThrows<IllegalArgumentException> {
            runBlocking {
                service.verifyCredentials(request)
            }
        }
    }

    @Test
    fun testVerifyCredentialsWithQueryFailure() {
        val request = verifyCredentialsRequest { email = "joe@gmail.com"; password = "passwd" }
        every { validator.validateGrpcRequest_VerifyCredentials(any()) } just Runs
        coEvery { userService.verifyCredentials(any(), any()) } returns Result.failure(Exception())

        assertThrows<Exception> {
            runBlocking {
                service.verifyCredentials(request)
            }
        }
    }

    @Test
    fun testVerifyCredentials() {
        val request = verifyCredentialsRequest { email = "joe@gmail.com"; password = "passwd" }
        every { validator.validateGrpcRequest_VerifyCredentials(any()) } just Runs

        val userid = UUID.randomUUID()
        val credentials = userCredentials("joe@gmail.com", HashUtils.hashPassword("passwd"), userid)
        coEvery { userService.verifyCredentials(any(), any()) } returns Result.success(credentials)
        val response = verifyCredentialsResponse {
            userId = uuidToUuid(userid)
        }
        every { mapper.mapResponseVerifyCredentials(any()) } returns response

        val result = runBlocking {
            service.verifyCredentials(request)
        }

        assertEquals(userid.toString(), result.userId.value)
    }

    @Test
    fun testGetUserProfileWithValidationFailure() {
        val request = getUserProfileRequest {}
        every { validator.validateGrpcRequest_getUserProfile(any()) } throws IllegalArgumentException()
        assertThrows<IllegalArgumentException> {
            runBlocking {
                service.getUserProfile(request)
            }
        }
    }

    @Test
    fun testGetUserProfileWithQueryFailure() {
        val userId = UUID.randomUUID()
        val request = getUserProfileRequest {
            userIds.add(uuidToUuid(userId))
        }
        every { validator.validateGrpcRequest_getUserProfile(any()) } just Runs
        coEvery {
            userService.getUserProfile(any())
        } returns Result.failure(Exception())

        assertThrows<Exception> {
            runBlocking {
                service.getUserProfile(request)
            }
        }
    }

    @Test
    fun testGetUserProfile() {
        val userid = UUID.randomUUID()
        val request = getUserProfileRequest {
            userIds.add(uuidToUuid(userid))
        }

        every { validator.validateGrpcRequest_getUserProfile(any()) } just Runs
        val user = mockk<User>()
        val users = listOf(user)
        val profile = mockk<UserProfile>()
        val response = getUserProfileResponse {
            profiles.add(profile)
        }
        every { mapper.buildGetUserProfileResponse(users) } returns response
        coEvery {
            userService.getUserProfile(any())
        } returns Result.success(users)
        val result = runBlocking {
            service.getUserProfile(request)
        }
        assertEquals(1, result.profilesCount)
    }

    private fun userCredentials(email: String, passwd: String, userid: UUID): UserCredentials {
        val user = mockk<User>()
        every { user.email } returns email
        every { user.userid } returns userid
        return UserCredentials.from(user, passwd)
    }
}
